controllers/applicationvolumereplication_controller.go

```diff
@@ -195,658 +276,895 @@ func (r *ApplicationVolumeReplicationReconciler) Reconcile(ctx context.Context,
+	avr *rmn.ApplicationVolumeReplication) (*plrv1.PlacementRule, *plrv1.PlacementRule, error) {
+	userPlRule, err := r.getUserPlacementRule(ctx, avr)
+	if err != nil {
+		return nil, nil, err
+	}
 
-	logger.Info(fmt.Sprintf("Processed %d/%d subscriptions. AVR Status Decisions %d. Requeue? %v",
-		len(a.instance.Status.Decisions), len(subscriptionList.Items), len(avr.Status.Decisions), requeue))
+	if err = r.annotatePlacementRule(ctx, avr, userPlRule); err != nil {
+		return nil, nil, err
+	}
 
-	done := !requeue
-	r.Callback(a.instance.Name, done)
+	avrPlRule, err := r.getOrClonePlacementRule(ctx, avr, userPlRule)
+	if err != nil {
+		return nil, nil, err
+	}
 
-	return ctrl.Result{Requeue: requeue}, nil
+	return avrPlRule, userPlRule, nil
 }
 
-type AVRInstance struct {
-	reconciler       *ApplicationVolumeReplicationReconciler
-	ctx              context.Context
-	log              logr.Logger
-	instance         *rmn.ApplicationVolumeReplication
-	subscriptionList *subv1.SubscriptionList
-}
-
-// For each subscription
-//		Check if it is paused for failover
-//			- restore PVs to the failed over cluster
-// 			- unpause
-//          - go to next subscription
-//		otherwise, select placement decisions
-//			- extract home cluster from placementrule.status.decisions
-//			- extract peer cluster from the clusters forming the dr pair
-//				example: ManagedCluster Set {A, B, C, D}
-//						 Pl.GenericPlacementField results in DR_Set = {A, B}
-//						 plRule{Status.Decision=A}
-//						 homeCluster = A
-//						 peerCluster = (DR_Set - A) = B
-//		create or update ManifestWork
-// returns placement decisions which can be the decisions for only a subset of subscriptions
-//
-func (a *AVRInstance) processSubscriptions() bool {
-	a.log.Info("Process subscriptions", "total", len(a.subscriptionList.Items))
+func (r *ApplicationVolumeReplicationReconciler) getUserPlacementRule(ctx context.Context,
+	avr *rmn.ApplicationVolumeReplication) (*plrv1.PlacementRule, error) {
+	r.Log.Info("Getting User PlacementRule", "placement", avr.Spec.Placement)
 
-	requeue := false
+	if avr.Spec.Placement == nil || avr.Spec.Placement.PlacementRef == nil {
+		return nil, fmt.Errorf("invalid user placementRule for AVR %s", avr.Name)
+	}
 
-	for idx, subscription := range a.subscriptionList.Items {
-		// On the hub ignore any managed cluster subscriptions, as the hub maybe a managed cluster itself.
-		// SubscriptionSubscribed means this subscription is child sitting in managed cluster
-		// Placement.Local is true for a local subscription, and can be used in the absence of Status
-		if subscription.Status.Phase == subv1.SubscriptionSubscribed ||
-			(subscription.Spec.Placement != nil && subscription.Spec.Placement.Local != nil &&
-				*subscription.Spec.Placement.Local) {
-			a.log.Info("Skipping local subscription", "name", subscription.Name)
+	plRef := avr.Spec.Placement.PlacementRef
 
-			continue
-		}
+	if plRef.Namespace == "" {
```

Comment:
All PlRule resources should have a namespace (even `default`), this condition bothers me.
PlRule CRD is a namespaced CRD, so why should this go missing? The case could be that the user did not provide one in AVR spec, and if so can we add a log line here stating we are assuming the namespace for the PlRule?